pkg/disk/group_volume_snapshot_utils.go

```diff
@@ -0,0 +1,259 @@
+	}
+	creationTime := timestamp.Timestamp{Seconds: t.Unix()}
+
+	readyToUse := groupSnapshot.Status == SnapshotStatusAccomplished
+	snapshots := []*csi.Snapshot{}
+	for _, ecsSnapshot := range groupSnapshot.Snapshots.Snapshot {
+		snapshot, err := formatCSISnapshot(&ecsSnapshot)
+		if err != nil {
+			return nil, err
+		}
+		snapshot.CreationTime = &creationTime
+		snapshot.GroupSnapshotId = groupSnapshot.SnapshotGroupId
+		if !snapshot.ReadyToUse {
+			// set readyToUse for groupsnapshots according to each snapshot status
+			readyToUse = false
+		}
+		snapshots = append(snapshots, snapshot)
+	}
+
+	return &csi.VolumeGroupSnapshot{
+		GroupSnapshotId: groupSnapshot.SnapshotGroupId,
+		Snapshots:       snapshots,
+		CreationTime:    &creationTime,
+		ReadyToUse:      readyToUse,
+	}, nil
+}
+
+// ifExistsGroupSnapshotMatchSourceVolume checks whether there SourceDiskId in `existsGroupSnapshot`
+// that correspond to all related snapshots in `sourceVolumeIds`.
+func ifExistsGroupSnapshotMatchSourceVolume(existsGroupSnapshot *ecs.SnapshotGroup, sourceVolumeIds []string) bool {
+	if len(existsGroupSnapshot.Snapshots.Snapshot) != len(sourceVolumeIds) {
+		return false
+	}
+	existsSnapshotSourceIdsMap := make(map[string]struct{})
+	for _, snapshot := range existsGroupSnapshot.Snapshots.Snapshot {
+		existsSnapshotSourceIdsMap[snapshot.SourceDiskId] = struct{}{}
+	}
+	matach := true
+	for _, id := range sourceVolumeIds {
+		if _, ok := existsSnapshotSourceIdsMap[id]; !ok {
+			matach = false
+			break
+		}
+	}
+	return matach
+}
+
+// ifExistsGroupSnapshotMatch checks whether there snapshotIDs in `snapshotIds`
+// that correspond to all related snapshots in `existsGroupSnapshot`.
+// If `requireExactMatch` is true, the function requires an exact match between
+// the related snapshotIDs and `snapshotIds`, including the same number of elements.
+// If `requireExactMatch` is false, it only checks if each related snapshotID
+// can be found in the `snapshotIds`, that means actual snapshots in group can be a subset of `snapshotIds`
+func ifExistsGroupSnapshotMatch(existsGroupSnapshot *ecs.SnapshotGroup, snapshotIds []string, requireExactMatch bool) bool {
+	if requireExactMatch && len(snapshotIds) != len(existsGroupSnapshot.Snapshots.Snapshot) {
+		return false
+	}
+	sourceSnapshotIdMap := make(map[string]struct{})
+	for _, id := range snapshotIds {
+		sourceSnapshotIdMap[id] = struct{}{}
+	}
+	matach := true
+	for _, snapshot := range existsGroupSnapshot.Snapshots.Snapshot {
+		if _, ok := sourceSnapshotIdMap[snapshot.SnapshotId]; !ok {
+			matach = false
+			break
+		}
+	}
+	return matach
+}
+
+// checkSourceVolumes checks whether a single source volume is valid.
+func checkSourceVolume(disk *ecs.Disk, zoneId string, capacityInGiB int) (
+	nZoneId string, nCapacityInGiB int, err error) {
+	nZoneId = zoneId
+	nCapacityInGiB = capacityInGiB
+
+	switch disk.Category {
+	case string(DiskESSD), string(DiskESSDAuto):
```

Comment:
Don't use switch, add a new field to CategoryDesc instead. So that we will not forget this when we add more disk type.